File: day09/streams.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);

while (<>) {
    # remove newline
    chomp;

    # remove cancels
    s/!.//g;

    # remove garbage
    my $garbage_len = 0;
    $garbage_len += length $1 while s/<([^>]*)>//;

    # compute score
    my @c = split //;
    my $score = 0;
    my $level = 0;

    for my $c (@c) {
        if ($c eq '{') {
            $level++;
        } elsif ($c eq '}') {
            $score += $level;
            $level--;
        }
    }
    say "score = $score";
    say "garbage length = $garbage_len";
}

File: day09/streams.py
#!/usr/bin/env python3
from sys import argv
import re

p = re.compile('<([^>]*)>')
filename = argv[1]
with open(filename) as f:
    for line in f:
        # remove newline
        line = line.rstrip("\n")
        
        # remove cancels
        line = re.sub('!.', '', line)

        # remove garbage
        garbage_len = 0
        m = p.search(line)
        while m is not None:
            garbage_len += len(m.group(1))
            line = p.sub('', line, count=1)
            m = p.search(line)

        # compute score
        score = 0
        level = 0
        for c in line:
            if c == '{':
                level += 1
            elif c == '}':
                score += level
                level -= 1

        print("score =", score)
        print("garbage length =", garbage_len)
        
